### DISCUSSION TITLE
Updateready

### DISCUSSION THREAD

**Discussion by ciss84** [aldostools/webMAN-MOD#1134] [General]
hi

**Comment by ciss84** [comment]
im need add updateready on this code

**Comment by ciss84** [comment]
<html manifest="84ciss.manifest"><head>
<title>ESP32-S2 Auto Cache(FW 9.00)</title>
<link rel="stylesheet prefetch" href="Style.css">
<script>history.replaceState(null,null,'index.html');</script>
</head>

	<body>		
        <script>
                if(window.applicationCache.status==2){localStorage.is755BCached=true;}
        	window.applicationCache.ondownloading=function(){document.getElementById("progress").innerHTML="Page Caching Started!!";localStorage.is755BCached=false;};
		window.applicationCache.onprogress=function(a){document.getElementById("myBar").style.width=(Math.round(100*(a.loaded/a.total)))+"%";document.getElementById("myBar").innerHTML=(Math.round(100*(a.loaded/a.total)))+"%";};
		window.applicationCache.oncached=function(){document.getElementById("progress").innerHTML="Page Cached Successfully!!";localStorage.is755BCached=true;setTimeout(function(){window.location.reload("myBar"); }, 1500);};
        	window.applicationCache.onnoupdate=function(){window.location.replace("lndex.html");};
        	window.applicationCache.onerror=function(){window.location.replace("lndex.html");};
        </script>
			<script>
			var width = 0;
			function frame1() {
				var elem = document.getElementById("myBar");
				var i=0;
				var id = setInterval(fra, 10);
				function fra(){
				  if(i<10 && width<100){
					  width++;
					  elem.style.width = width + "%";
					  elem.innerHTML = width  + "%";
				  } i++;
				}
			}
		</script>	
		<h1 id="progress" class="titlehead">Loading.. Please Wait !!</h1><hr><br>
		<h1 id="myProgress">
		  <div id="myBar">0%</div>
		</h1>
<div style="background-color:#032140;color:white;position:fixed;left:0;bottom:-0px;width:100%;text-align:center;">
<font style='color:#87FF33'>Modder & Dev : </font>Son84Goku - Find Me In Tweeter if you have any kind of issues !&nbsp;<a href="https://twitter.com/84Ciss" style='font-size:20px;text-align:center;color:yellow;'>@Son84Goku</a>
<br><font style='color:#87FF33'>Special thanks : </font>*Many Many Thanks to all Contributors in PS4 Scene ∆❍X❑*
</div>
</body>
</html>

**Comment by ciss84** [comment]
because cache updates block at 100% and do not display full cache !! first time works fine just update cache not works need reload page

**Comment by aldostools** [comment]
Hello ciss84
I have to go out now, but I will check your code when I'm back.

**Comment by aldostools** [comment]
Meanwhile, here is some suggestion by Google's Gemini:

The `updateready` event is a crucial part of the HTML5 Application Cache API, signaling that a new version of your application's cached resources is ready to be activated. This event is triggered when the browser has successfully downloaded the updated files specified in your application cache manifest file.

# How to Handle `updateready` Events

To handle the `updateready` event, you'll typically use JavaScript to listen for the event and then trigger the cache swap:

```JavaScript
window.addEventListener('load', function() {
  window.applicationCache.addEventListener('updateready', function() {
    if (window.applicationCache.status === window.applicationCache.UPDATEREADY)   
 {
      // Trigger the update
      window.applicationCache.swapCache();
      // Optionally, you can display a notification to the user
      alert('A new version of the app is available. Reloading now.');
      window.location.reload();
    }
  }, false);
});
```

## Breakdown of the Code:

1. Event Listener: We attach an event listener to the window.applicationCache object to listen for the updateready event.
2. Checking the Status: We check if the status property of the applicationCache object is equal to UPDATEREADY. This confirms that a new cache is ready.
3. Swapping the Cache: If the status is UPDATEREADY, we call the swapCache() method to activate the new cache.
4. User Notification: Optionally, we can display a notification to the user, informing them about the update and reloading the page.

## Key Points to Remember:

- Manifest File: The updateready event is triggered based on changes in your application cache manifest file. Ensure that you update the manifest file to include new or modified resources.
- Browser Compatibility: While the Application Cache API is widely supported, there are some browser-specific quirks and limitations. It's essential to test your application across different browsers.
- User Experience: Consider the user experience when handling updates. You might want to provide a graceful way to inform users about the update and allow them to choose when to apply it.

## Modern Alternatives to Application Cache

While the Application Cache API was a useful tool in its time, modern web development practices often favor more flexible and efficient approaches:

- Service Workers: Service Workers provide a more powerful and flexible way to implement offline capabilities and caching strategies.
- HTTP Caching: Leveraging HTTP caching headers can be effective for caching static assets.

By understanding the updateready event and exploring modern alternatives, you can effectively manage the update process and enhance the user experience of your web applications.


Sources and related content

**Comment by aldostools** [comment]
![image](https://github.com/user-attachments/assets/70d9e2ac-f9a9-46a8-a9b9-9a00af38ca7d)

**Comment by ciss84** [comment]
possible you update this code ?? have add and not works on ps5 just blok !! I would like to understand my mistake !! when you have time

**Reply by aldostools** [reply]
I will check if I can find something when I get back home. But I cannot promise you much, since I don't have a proper environment to test the code.

**Reply by ciss84** [reply]
sent me on you have time im test tomorow im go sleep !! big thanks

**Comment by aldostools** [comment]
I noticed that you have lndex.html (first letter is an L in lowercase) in the update and error events instead of index.html 

Is that right?

**Reply by ciss84** [reply]
yes lndex is second html

**Reply by ciss84** [reply]
after first cache lndex.html is loaded !! just after update cache block

**Reply by ciss84** [reply]
on ps4 possible to clean the cache and works!! ps5 problem not working clean web cache!! for this I need the update cache

**Reply by aldostools** [reply]
The function frame1() is never called.
It seems that the script block is unused.

Note: I'm reading the code from a cellphone. It's difficult to analyze code. I'll continue at home.

**Comment by aldostools** [comment]
Checking the code used by the project [ps5xploit](https://github.com/aldostools/ps5xploit)

It uses a [cache manifest](https://github.com/aldostools/ps5xploit/blob/main/cache.appcache) file loaded in: `<html manifest="cache.appcache">`

The `updateready` event is handle in the javascript [appcache_handler.js](https://github.com/aldostools/ps5xploit/blob/main/js/appcache_handler.js#L69-L73)
````javascript
    var appCache = window.applicationCache;

    appCache.addEventListener('updateready', function (e) {
        if (window.applicationCache.status == window.applicationCache.UPDATEREADY) {
            showToast('The site was updated. Refresh to switch to updated version',8000);
        }
    }, false);
````

The function showToast() used to show messages is found in [custom_host_stuff.js](https://github.com/aldostools/ps5xploit/blob/main/js/custom_host_stuff.js#L381-L400)
It uses the `toast` css style from [main.css](https://github.com/aldostools/ps5xploit/blob/main/main.css#L244-L274)

**Reply by ciss84** [reply]
oki im look tnx

**Comment by ciss84** [comment]
tnx have fixed

<html manifest="84ciss.manifest"><head>
<title>Auto Cache(FW 5.50)</title>
<link rel="stylesheet prefetch" href="Style.css">
<script>history.replaceState(null,null,'index.html');</script>
</head>

	<body>		
        <script>
                if(window.applicationCache.status==2){localStorage.is755BCached=true;}
        	window.applicationCache.ondownloading=function(){document.getElementById("progress").innerHTML="Page Caching Started!!";localStorage.is755BCached=false;};
		window.applicationCache.onprogress=function(a){document.getElementById("myBar").style.width=(Math.round(100*(a.loaded/a.total)))+"%";document.getElementById("myBar").innerHTML=(Math.round(100*(a.loaded/a.total)))+"%";};
		window.applicationCache.oncached=function(){document.getElementById("progress").innerHTML="Page Cached Successfully!!";localStorage.is755BCached=true;setTimeout(function(){window.location.reload("myBar"); }, 1500);};
            window.applicationCache.onupdateready=function(){document.getElementById("progress").innerHTML="Page Cached Successfully!!";localStorage.is755BCached=true;setTimeout(function(){window.location.reload("myBar"); }, 1500);};
        	window.applicationCache.onnoupdate=function(){window.location.replace("lndex.html");};
        	window.applicationCache.onerror=function(){window.location.replace("lndex.html");};
        </script>
			<script>
			var width = 0;
			function frame1() {
				var elem = document.getElementById("myBar");
				var i=0;
				var id = setInterval(fra, 10);
				function fra(){
				  if(i<10 && width<100){
					  width++;
					  elem.style.width = width + "%";
					  elem.innerHTML = width  + "%";
				  } i++;
				}
			}
		</script>	
		<h1 id="progress" class="titlehead">Loading.. Please Wait !!</h1><hr><br>
		<h1 id="myProgress">
		  <div id="myBar">0%</div>
		</h1>
    <div style="background-color:#043836;color:white;position:fixed;left:0;bottom:-0px;width:100%;text-align:center;">
    <font style='color:#87FF33'>Modder & Dev : </font>84Ciss - Find Me In Tweeter if you have any kind of issues !&nbsp;<a href="https://twitter.com/84Ciss" style='font-size:20px;text-align:center;color:yellow;'>@84Ciss</a>
    <br><font style='color:#87FF33'>Special thanks : </font>*Many Many Thanks to all Contributors in PS4/PS5 Scene ∆❍X❑*
</div>
</body>
</html>
